File: 0110_balanced_binary_tree/is_balanced.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Time: O(n)
 * Space: O(n)
 */
func isBalanced(root *TreeNode) bool {
	return dfs(root) != -1
}

func dfs(node *TreeNode) int {
	if node == nil {
		return 0
	}
	left := dfs(node.Left)
	if left == -1 {
		return -1
	}
	right := dfs(node.Right)
	if right == -1 {
		return -1
	}
	if abs(left-right) > 1 {
		return -1
	}
	return max(left, right) + 1
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func max(x, y int) int {
	if x >= y {
		return x
	}
	return y
}

func main() {

}

File: 0110_balanced_binary_tree/is_balanced.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def isBalanced(root: TreeNode) -> bool:
    """
    Time: O(n)
    Space: O(n)
    """
    return dfs(root) != -1
    
def dfs(node: TreeNode) -> int:
    if not node:
        return 0
    left = dfs(node.left)
    if left == -1:
        return -1
    right = dfs(node.right)
    if right == -1:
        return -1
    if abs(left - right) > 1:
        return -1
    return max(left, right) + 1